### ISSUE TITLE
Get rid of liftMay

### ISSUE THREAD

**Issue by treeowl** [ndmitchell/safe#12]
`liftMay` is convenient, but it has two major flaws:

1. Every use must be considered trusted code, since it wraps an unsafe function.
2. It interferes with list fusion optimizations.

So I'd love to kill it.

**Comment by ndmitchell** [comment]
`liftMay` is an internal function, so killing it is not infeasible - the question is what to replace it with?

1. Every use must be trusted, but we're taking some unsafe functions and trusting them, so I think somewhere has to introduce the trust.
2. Does returning a Maybe fundamentally interfere with list fusion? Is there some reason liftMay makes it particularly worse?

**Comment by treeowl** [comment]
The alternative is to reimplement functions rather than wrap them. Take a
look at the default definition of foldr1 in Data.Foldable. You'll see that
it's just foldr1May wrapped up to make it unsafe!

liftMay completely blocks list fusion. On the argument side, or does so by
inspecting the argument before folding over it. On the result side, it
(when relevant) produces a Maybe [a] instead of a [a], so it can't fuse
with a fold.

**Comment by ndmitchell** [comment]
True, for foldr1 going via foldr1May is definitely a better approach in every way.

And yep, in some cases rewriting is probably better than the intermediate liftMay. Of course, rewriting means more testing is required - important to check the behaviour is the same.

Although I don't think I'd say liftMay is never the appropriate choice - I think it's not a bad fallback if none of the other approaches is better.

**Comment by treeowl** [comment]
For lists, I don't *think* it should ever be needed. For Foldable, the
story is potentially more complex, since an instance could offer an
optimized foldr1 or foldl1.

BTW, here are another couple (initSafe can probably be made even more
fusion friendly by using GHC's `build`, but it's way past my bedtime):

initSafe :: [a] -> [a]
initSafe xs = foldr go (`seq` []) xs Nothing where
  go x r Nothing = r (Just x)
  go x r (Just p) = p : r (Just x)

initMay :: [a] -> Maybe [a]
initMay xs = foldr go Nothing xs where
  go x rest = Just $
    case rest of
      Nothing -> []
      Just next -> x : next

Incidentally, I suspect it may be possible to implement initMay more
elegantly with the help of a hypothetical foldMap1May, but I haven't tried
yet.

**Comment by treeowl** [comment]
Hmm... I may have gotten too fancy with initMay; Foldable on the brain.
It's likely better to go with a simpler approach there.

initMay [] = Nothing
initMay (x : xs) = Just (go x xs) where
  go _ [] = []
  go a (b : bs) = a : go b bs

**Comment by treeowl** [comment]
BTW, the ultimate fusion-friendly initSafe will (modulo various RULES and
such) look something like this:

initSafe :: [a] -> [a]
initSafe xs = build $ \c n ->
  let
    go x r Nothing = r (Just x)
    go x r (Just p) = p `c` r (Just x)
  in foldr go (`seq` n) xs Nothing

**Comment by ndmitchell** [comment]
What's the `seq`? Is:

    initSafe [] = []
    initSafe xs = init xs

Not a viable approach? That way whatever rules are associated with `init` will fire, and we guarantee identical strictness properties etc.

**Comment by treeowl** [comment]
No, not viable, because GHC will avoid the loss of sharing. I also prefer
to avoid using partial functions unless it's necessary for speed.

**Comment by treeowl** [comment]
Oh, and the seq is to aid strictness analysis. It's just forcing the
outside of the Maybe that's passing through from left to right, so there's
nothing shady about it. Same as `\m -> maybe [] (const []) m`, but clearer.

**Comment by treeowl** [comment]
Sorry, should have been more explicit. GHC fears duplicating work by firing
rules. Essentially, if it sees you're looking at the list twice, it figures
you really want the list, so it gives it to you. No fusion.

**Comment by ndmitchell** [comment]
Sure, if tests and examination of the Core lead you to that definition, who am I to argue!

**Comment by treeowl** [comment]
I haven't done any perf testing or core examination today. Just working on
what I've seen and read. I'll do the grunt work later, as I've done it in
the past.
